Modelica/Mechanics/MultiBody/Frames/Quaternions/from_T.mo

```diff
@@ -10,63 +10,42 @@ function from_T
   output Quaternions.Orientation Q
     "Quaternions orientation object to rotate frame 1 into frame 2 (Q and -Q have same transformation matrix)";
 protected
-  Real paux;
-  Real paux4;
-  Real c1;
-  Real c2;
-  Real c3;
-  Real c4;
-  constant Real p4limit=0.1;
-  constant Real c4limit=4*p4limit*p4limit;
+  Real t;
 algorithm
   /*
-   Note, for quaternions, Q and -Q have the same transformation matrix.
-   Calculation of quaternions from transformation matrix T:
-   It is guaranteed that c1>=0, c2>=0, c3>=0, c4>=0 and
-   that not all of them can be zero at the same time
-   (e.g., if 3 of them are zero, the 4th variable is 1).
-   Since the sqrt(..) has to be performed on one of these variables,
-   it is applied on a variable which is far enough from zero.
-   This guarantees that the sqrt(..) is never taken near zero
-   and therefore the derivative of sqrt(..) can never be infinity.
+   This is an optimized version of the quaternion conversion based on
+   this paper: https://d3cw3dd2w32x2b.cloudfront.net/wp-content/uploads/2015/01/matrix-to-quat.pdf
+*/
+  if T[3, 3] < 0 then
+    if T[1, 1] > T[2, 2] then
+      t := 1 + T[1, 1] - T[2, 2] - T[3, 3];
+      Q := {t, T[1, 2] + T[2, 1], T[3, 1] + T[1, 3], T[2, 3] - T[3, 2]};
+    else
+      t := 1 - T[1, 1] + T[2, 2] - T[3, 3];
+      Q := {T[1, 2] + T[2, 1], t, T[2, 3] + T[3, 2], T[3, 1] - T[1, 3]};
+    end if;
+  else
+    if T[1, 1] < -T[2, 2] then
+      t := 1 - T[1, 1] - T[2, 2] + T[3, 3];
+      Q := {T[3, 1] + T[1, 3], T[2, 3] + T[3, 2], t, T[1, 2] - T[2, 1]};
+    else
+      t := 1 + T[1, 1] + T[2, 2] + T[3, 3];
+      Q := {T[2, 3] - T[3, 2], T[3, 1] - T[1, 3], T[1, 2] - T[2, 1], t};
+    end if;
+  end if;
+  Q := Q .* 0.5/sqrt(t);
```

Comment:
```suggestion
  Q := Q*0.5/sqrt(t);
```
The component-wise multiplication is not relevant when multiplying by a scalar and made me wondering first.